components/diagram.jsx

```diff
@@ -1,20 +1,16 @@
-import log from 'loglevel';
+import { useRef, useEffect } from "react";
+import log from "loglevel";
 
-export class Diagram extends React.Component {
+export const Diagram = ({ connections }) => {
+  const canvas = useRef();
+  const container = useRef();
 
-  constructor(props) {
-    super(props)
-    this.state = {}
-  }
+  const updateCanvas = () => {
+    if (!canvas) return;
+    if (!canvas.current) return;
 
-  setConnections(connections) {
```

Comment:
connections state has been moved out, into its own hook called useKeyConnections. Moving the state out allows other components to access it